core/src/main/scala/org/apache/spark/deploy/security/KafkaDelegationTokenProvider.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.security
+
+import scala.reflect.runtime.universe
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.security.Credentials
+import org.apache.hadoop.security.token.{Token, TokenIdentifier}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.{KAFKA_BOOTSTRAP_SERVERS, KAFKA_SECURITY_PROTOCOL}
+import org.apache.spark.util.Utils
+
+private[security] class KafkaDelegationTokenProvider
```

Comment:
Another alternative that may be cleaner than this is to actually have this class in the spark-sql-kafka module, and instantiate it in `HadoopDelegationTokenManager` using `Utils.classForName(...).newInstance()`.

Then this particular code does not need to use reflection at all, and you may be able to clean some code up on the `TokenUtils` side.

Reply:
Checking it...

Reply:
That will make `KafkaDelegationTokenProvider` more clean but has a couple of drawbacks:
 * Just move the reflection into `HadoopDelegationTokenManager`
 * Has to be copied when delegation token will be introduced in DStreams

Reply:
> Just moves the reflection into HadoopDelegationTokenManager

You can't escape reflection unless you use `java.util.ServiceLoader` here, and it was decided not to use that when this code was first written. My argument is that it is a lot less reflection - it's just `Class.forName` instead of this whole class here.

> Has to be copied when delegation token will be introduced in DStreams
 
And how is that different from what you have here? The code you have won't work for dstreams since it's not even available when you pull in just the dstream connector.

Reply:
In general if we're not moving this class into kafka area the only change which has to be done is to add another package from where the `TokenUtil` can be loaded. Otherwise `KafkaDelegationTokenProvider` has to be copied from `spark-sql-kafka...` to `spark-streaming-kafka...`. I think leaving as it is makes it more simple.

Reply:
Using full-reflection would make the code hell complicated but your other suggestion about provided scope is really good. This way the copy-paste can be avoided. Changed accordingly.